### PR TITLE
Read from an idle blaze-client connection to detect when server closes it

### LINKED ISSUE #4798: Blaze Client keeps half-closed connection after server sends FIN-ACK through an idle connection
This is a problem because both sides can't release resources allocated to the connection, but also, because blaze will later try and fail to use that connection.

I prepared code that reproduces this behaviour using `BlazeServer`'s `idleTimeout` as the mechanism closing the connection on server's side. 

The scenario:
- make a request 
- wait until server closes the connection (due to 5s idle timeout)
- make another request

For comparison there's the same scenario but using `AsyncHttpClient`. The behaviour of `AsyncHttpClient` is in my opinion the expected behaviour (It's also consistent with for example Firefox's behaviour).

```
  val server = BlazeServerBuilder[IO](munitExecutionContext)
    .withIdleTimeout(5.second)
    .bindLocal(12345)
    .withHttpApp(Router("/" -> HttpRoutes.of[IO] { case GET -> Root => IO.sleep(1.seconds) >> Ok("") }).orNotFound)
    .resource

  test("Blaze Client keeps half-closed connection after server sends FIN-ACK through an idle connection") {
    (
      server,
      BlazeClientBuilder[IO](munitExecutionContext).resource
    ).tupled.use { case (_, client) =>for {
      _ <- client.expect[String](Uri.fromString(s"http://127.0.0.1:12345").yolo).attemptTap(result => IO(println(result)))
      _ <- IO.sleep(6.seconds)
      _ <- client.expect[String](Uri.fromString(s"http://127.0.0.1:12345").yolo).attemptTap(result => IO(println(result)))
    } yield ()
    }.unsafeRunSync()
  }

  test("AsyncHttpClient closes connection after server sends FIN-ACK through an idle connection") {
    (
      server,
      AsyncHttpClient.resource[IO]()
    ).tupled.use { case (_, client) => for {
      _ <- client.expect[String](Uri.fromString(s"http://127.0.0.1:12345").yolo).attemptTap(result => IO(println(result)))
      _ <- IO.sleep(6.seconds)
      _ <- client.expect[String](Uri.fromString(s"http://127.0.0.1:12345").yolo).attemptTap(result => IO(println(result)))
    } yield ()
    }.unsafeRunSync()
  }
```
(Full runnable code: https://github.com/RafalSumislawski/http4s/tree/blaze-half-closed-connection-issue)

The code itself just reproduces the behaviour, without checking if it's correct. I use wireshark to analyse to communication between the client and server.

### Blaze client:
image

In packet 1519 server closes the connection. The client ACKs it, but it doesn't send a FIN, so the connection remains half-closed (CLOSE_WAIT from server's perspective). Then packet 1521 is clients attempt to use that connection to send another request, to which the server responds with RST finally killing the connection. The client then establishes a new connection and successfully resends the request. Nonetheless precious resources have been wasted on the attempt to send request though closed connection. The retry mechanism was useful in this scenario, but it just covers the problem, not fixes it. And I see some other scenarios where it causes problems.

### AsyncHttpClient:
image

After receiving FIN-ACK, the AsyncHttpClient immediately responds with FIN-ACK releasing resource on both ends. When a new request needs to be sent, I understands that it needs to open a new connection.

I realise this may be an issue inside blaze, not http4s, nonetheless I think the perspective of http4s-client user, and the comparison with other implementation is useful for clear definition of the problem and the expected behaviour.

### PR DESCRIPTION
**Pull request by RafalSumislawski**
Fixes #4798

### CODE REVIEW

**Comment by RafalSumislawski** [comment]
Is it just me or did 0.21.11 disappear from maven repo causing these mima failures?

File: blaze-client/src/main/scala/org/http4s/client/blaze/Http1Connection.scala [hunk 1]
```diff
@@ -107,9 +111,9 @@ private final class Http1Connection[F[_]](
   def resetRead(): Unit = {
     val state = stageState.get()
     val nextState = state match {
-      case Idle => Some(Idle)
+      case i: Idle => Some(i)
       case ReadWrite => Some(Write)
-      case Read => Some(Idle)
+      case Read => Some(Idle(Some(startIdleRead())))
```
Comment (RaasAhsan) [inline blaze-client/src/main/scala/org/http4s/client/blaze/Http1Connection.scala:116]: So the idle read will begin immediately after a connection is set to idle after being used ? Regardless of whether it is acquired again

Comment (RafalSumislawski) [reply blaze-client/src/main/scala/org/http4s/client/blaze/Http1Connection.scala:116]: Exactly. This way we keep reading while the connection sits idle in the pool. If we happen to read an EOF, the connection gets closed. Once the connection is borrower and used, the read operation that we started here is used as the first operation in the process of reading a response.

**Comment by RafalSumislawski** [comment]
I should have mentioned, that this PR doesn't touch the `PoolManager`, because logic preventing borrowing closed connections is already in place: https://github.com/http4s/http4s/blob/bd9a64d95a9d280f46fa0ce3ebd80f741ef96bb0/client/src/main/scala/org/http4s/client/PoolManager.scala#L181

**Comment by RafalSumislawski** [comment]
This PR makes issues similar to #2277 appear much more often. Please consider merging #4815 before this PR.

File: blaze-client/src/main/scala/org/http4s/client/blaze/Http1Connection.scala [hunk 2]
```diff
@@ -380,7 +416,7 @@ private object Http1Connection {
 
   // ADT representing the state that the ClientStage can be in
   private sealed trait State
-  private case object Idle extends State
+  private final case class Idle(idleRead: Option[Future[ByteBuffer]]) extends State
```
Comment (rossabaker) [inline blaze-client/src/main/scala/org/http4s/client/blaze/Http1Connection.scala:419]: Is `Idle(None)` just an init state?  I might have made a separate `State` and eliminated the `Option`, but I guess we're treating both variants of `Idle` the same in lots of places.  Maybe my idea cleans it up, maybe it makes it worse.  I trust your judgment.

**Review by rossabaker — DISMISSED** [review]

File: blaze-client/src/main/scala/org/http4s/client/blaze/Http1Connection.scala (hunk 2 above)
Comment (RafalSumislawski) [reply blaze-client/src/main/scala/org/http4s/client/blaze/Http1Connection.scala:417]: Yes, it's just the state before running the first request. 

I just had a second look at that state and noticed I can remove two lines of code: ff0cc53 

I tried refactoring the code to have this separate state. This is what I came up with:
Option A: https://github.com/RafalSumislawski/http4s/commit/efb7afbb79eaeb252f0caabda0bb151921d6bc47
Option B: https://github.com/RafalSumislawski/http4s/commit/ef024654a38e0bb90cf998aab5a850897fbf129c
I like the fact that the additional state makes the model of the state machine more clear, but I don't like the additional LoC. Please let me know if I should bring one of these options into this PR. If I were to make the decision, I would choose the current shape of the PR.

Comment (rossabaker) [reply blaze-client/src/main/scala/org/http4s/client/blaze/Http1Connection.scala:417]: Thanks for trying.  I think you're right.  Let's keep it as is.

**Comment by rossabaker** [comment]
Build of ff0cc53 failed because git checkout was 500ing.  Restarting.

**Comment by rossabaker** [comment]
The MiMa failure is legitimate, because the case object became a case class.  I don't usually like exceptions, but this is private state, and should be fine.

**Comment by rossabaker** [comment]
I'm just going to do the MiMa work so we can move this and others along.

**Pull request merged**